components/scream/src/physics/zm/zm_inputs_initializer.cpp

```diff
@@ -0,0 +1,239 @@
+  set_grid_opts_helper(dsubcld, "dsubcld", true, NULL, LINEAR);
+  set_grid_opts_helper(jt, "jt", true, NULL, LINEAR);
+  set_grid_opts_helper(maxg, "maxg", true, NULL, LINEAR);
+  set_grid_opts_helper(ideep, "ideep", true, NULL, LINEAR);
+  set_grid_opts_helper(lengath, "lengath", true, NULL, SCALAR_3D_MID);
+  set_grid_opts_helper(ql, "ql", true, NULL, SCALAR_3D_MID);
+  set_grid_opts_helper(rliq, "rliq", true, NULL, SCALAR_3D_MID);
+  set_grid_opts_helper(landfrac, "landfrac", true, NULL, LINEAR);
+  set_grid_opts_helper(hu_nm1, "hu_nm1", true, NULL, SCALAR_3D_MID);
+  set_grid_opts_helper(cnv_nm1, "cnv_nm1", true, NULL, SCALAR_3D_MID);
+  set_grid_opts_helper(tm1, "tm1", true, NULL, SCALAR_3D_MID);
+  set_grid_opts_helper(qm1, "qm1", true, NULL, SCALAR_3D_MID);
+  set_grid_opts_helper(t_star, "t_star", true, NULL, SCALAR_3D_MID);
+  set_grid_opts_helper(q_star, "q_star", true, NULL, SCALAR_3D_MID);
+  set_grid_opts_helper(dcape, "dcape", true, NULL, LINEAR);
+  set_grid_opts_helper(q, "q", true, NULL, SCALAR_3D_MID);
+  set_grid_opts_helper(tend_s, "tend_s", true, NULL, LINEAR);
+  set_grid_opts_helper(tend_q, "tend_q", true, NULL, LINEAR);
+  set_grid_opts_helper(cld, "cld", true, NULL, LINEAR);
+  set_grid_opts_helper(snow, "snow", true, NULL, LINEAR);
+  set_grid_opts_helper(ntprprd, "ntprprd", true, NULL, SCALAR_3D_MID);
+  set_grid_opts_helper(ntsnprd, "ntsnprd", true, NULL, SCALAR_3D_MID);
+  set_grid_opts_helper(flxprec, "flxprec", true, NULL, SCALAR_3D_MID);
+  set_grid_opts_helper(flxsnow, "flxsnow", true, NULL, SCALAR_3D_MID);
+  set_grid_opts_helper(ztodt, "ztodt", true, NULL, SCALAR_3D_MID);
+  set_grid_opts_helper(pguall, "pguall", true, NULL, VECTOR_3D_MID);
+  set_grid_opts_helper(pgdall, "pgdall", true, NULL, VECTOR_3D_MID);
+  set_grid_opts_helper(icwu, "icwu", true, NULL, VECTOR_3D_MID);
+  set_grid_opts_helper(ncnst, "ncnst", true, NULL, VECTOR_3D_MID);
+  set_grid_opts_helper(fracis, "fracis", true, NULL, VECTOR_3D_MID);
+
+
+} 
+void ZMInputsInitializer::add_field (const field_type &f)
+{
+  const auto& id = f.get_header().get_identifier();
+  
+  m_fields.emplace(id.name(),f);
+  m_fields_id.insert(id);
+}
+
+
+
+void ZMInputsInitializer :: initialize_fields(){
+
+  //check if zm inputs have been registered as fields
+  int count = 0;
+  for (int j = 0; j < zm_inputs.size(); j++){
+    count += m_fields.count(zm_inputs[j]);
+    if (count==0){
+      return;
+    }
+  }
+  
+
+  scream_require_msg (count==INPUT_SIZE,
+    "Error! ZMInputsInitializer is expected to init _______________.\n"
+    "Only " + std::to_string(count) + " of those have been found.\n"
+    "Please, check the atmosphere processes you are using,"
+    "and make sure they agree on who's initializing each field.\n");
+
+
+
+  for (int i = 0; i < zm_inputs.size(); i++){
+    //Get and store device view using input name
+    Kokkos::View<Real*, Kokkos::LayoutRight, HostDevice> d_v = m_fields.at(zm_inputs[i]).get_view();
```

Comment:
Using the keyword `auto` might make this line and the next a bit less "heavy". Like:

```
auto d_v = ...
auto h_m = Kokkos::create_mirror_view(d_v);
```
Also, but this is somewhat a personal taste subject, it might maker the code easier if you used a pattern like "blah_d"/"blah_h" for device/host views (or even "blah_dev"/"blah_host"), or "d_blah"/"h_blah" if you like to prepend the memory location.

Reply:
I actually changed the auto convention intentionally, because the auto keyword made putting values in structs and vectors difficult if not impossible. I can change the structure so that it uses auto but I think that the cleanliness of auto would mean adding more lines of more muddled code